### PR TITLE
asynchronous apex features

### PR DESCRIPTION
**Pull request by Aggnieszkka**

### CODE REVIEW

File: InnovativeCars/force-app/main/default/classes/CarToShowroomHandler1.cls [hunk 1]
```diff
@@ -91,4 +91,22 @@ public class CarToShowroomHandler1 {
             }
```
Comment (sylwiachr) [inline InnovativeCars/force-app/main/default/classes/CarToShowroomHandler1.cls:88]: Why 1 at the end of the class name?

File: InnovativeCars/force-app/main/default/classes/NewOffersEmailSender.cls [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    * @date 06.2021
+    *
+    * @description method sending email about new offers
+    * @param newOffersIds list of new offers
+    */
+    @future (callout = true)
+    public static void sendEmail(List<ID> newOffersIds){
+
+        Messaging.SingleEmailMessage mail = new Messaging.SingleEmailMessage();
+        String[] toAddresses = new String[] {Constants.NEW_OFFERS_EMAIL_ADDRESS};
```
Comment (sylwiachr) [inline InnovativeCars/force-app/main/default/classes/NewOffersEmailSender.cls]: why email from constants?

File: InnovativeCars/force-app/main/default/classes/DeleteOldTelemetries.cls [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+* @author Agnieszka Ząbkowicz
+* @date 06.2021
+*
+* @description class for deleting telemetries older than 90 days
+*/
+public class DeleteOldTelemetries implements Database.Batchable<sObject> {
+
+    public Database.QueryLocator start(Database.BatchableContext bc) {
+        date d = system.today().adddays(-90);
+        List<Telemetry__c> tel = [SELECT Id FROM Telemetry__c WHERE DAY_ONLY(CreatedDate) <= :d];
```
Comment (sylwiachr) [inline InnovativeCars/force-app/main/default/classes/DeleteOldTelemetries.cls]: this line was only for the purpose of tests I guess?

File: InnovativeCars/force-app/main/default/classes/DeleteOldTelemetriesTest.cls [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+*
+* @description test class for DeleteOldTelemetries class
+*/
+@isTest
+public class DeleteOldTelemetriesTest {
+
+    /**
+    * @author Agnieszka Ząbkowicz
+    * @date 06.2021
+    *
```
Comment (sylwiachr) [inline InnovativeCars/force-app/main/default/classes/DeleteOldTelemetriesTest.cls]: why this empty line? please also add proper date

File: InnovativeCars/force-app/main/default/classes/DeleteOldTelemetriesTest.cls [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        for(Telemetry__c telemetry : telemetries){
+            Test.setCreatedDate(telemetry.Id, createdDate);
+        }
+
+        Map<Id, Telemetry__c> IdToTelemetryMap = new Map<Id, Telemetry__c>(telemetries);
+        List<Id> telemetriesIds = new List<Id>(IdToTelemetryMap.keySet());
+      
+        Test.startTest();
+            DeleteOldTelemetries deleteOld = new DeleteOldTelemetries();
+        	Id batchId = Database.executeBatch(deleteOld);
```
Comment (sylwiachr) [inline InnovativeCars/force-app/main/default/classes/DeleteOldTelemetriesTest.cls]: formatting

File: InnovativeCars/force-app/main/default/classes/DeleteOldTelemetriesTest.cls [hunk 6]
```diff
@@ -0,0 +1,53 @@
+/**
+* @author Agnieszka Ząbkowicz
+* @date 06.2021
+*
+* @description test class for DeleteOldTelemetries class
```
Comment (sylwiachr) [inline InnovativeCars/force-app/main/default/classes/DeleteOldTelemetriesTest.cls:4]: there should be separate class testing batch and separate one testing scheduler

File: InnovativeCars/force-app/main/default/classes/Constants.cls [hunk 7]
```diff
@@ -7,11 +7,23 @@ public class Constants {
...
     public static final String LESS_THAN_ZERO_ERROR = 'Number of car spaces cannot be less then 0!';
     public static final String LESS_THAN_CARS_AVAILABLE_ERROR = 'Number of car spaces cannot be less then number of cars available at the showroom!';
+    public static final String TELEMETRIES_DELETED_MESSAGE = 'Telemetries older than 90 days deleted successfully';
+    public static final String EMAIL_SEND_SUCCESSFULLY_MESSAGE = 'The email was sent successfully.';
+    public static final String EMAIL_SEND_FAILED_MESSAGE = 'The email failed to send: ';
+    public static final String NEW_OFFERS_EMAIL_ADDRESS = 'noviv37645@0ranges.com';
+    public static final String NEW_OFFERS_EMAIL_SUBJECT = 'New offers added!';
+    public static final String NEW_OFFERS_EMAIL_BODY = 'New offers have been added to showrooms. Number of offers: ';
+    public static final String CASE_SUBJECT = 'Renovation Case ';
+    public static final String CASE_ORIGIN = 'Web';
```
Comment (sylwiachr) [inline InnovativeCars/force-app/main/default/classes/Constants.cls]: the name is not meaningfull. the same as case subject. please consider something like CASE_ORIGIN_WEB (similarly with case subject)

File: InnovativeCars/force-app/main/default/classes/RenovationTeamRequestSender.cls [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+            newCases.add(new Case(
+                Subject = Constants.CASE_SUBJECT + showroom.Name, 
+                Showroom__c = showroom.Id, 
+                Origin = Constants.CASE_ORIGIN
+                ));
+        }
+
+        Http server = new Http();
+        HttpRequest req = new HttpRequest();
+        req.setEndpoint(Constants.RENOVATION_ENDPOINT);
```
Comment (sylwiachr) [inline InnovativeCars/force-app/main/default/classes/RenovationTeamRequestSender.cls:17]: consider having endpoint in named credentials / custom metadata / custom setting

File: InnovativeCars/force-app/main/default/classes/DeleteOldTelemetries.cls [hunk 9]
```diff
@@ -0,0 +1,20 @@
+/**
+* @author Agnieszka Ząbkowicz
+* @date 15.06.2021
+* @description class for deleting telemetries older than 90 days
+*/
+public class DeleteOldTelemetries implements Database.Batchable<sObject> {
+
+    public Database.QueryLocator start(Database.BatchableContext bc) {
+        date d = system.today().adddays(-90);
```
Comment (AdrianDlu) [inline InnovativeCars/force-app/main/default/classes/DeleteOldTelemetries.cls]: bad variable name d

File: InnovativeCars/force-app/main/default/classes/DeleteOldTelemetries.cls [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+        return Database.getQueryLocator(
+            'SELECT Id FROM Telemetry__c WHERE DAY_ONLY(CreatedDate) <= :d'
+        );
+    }
+    public void execute(Database.BatchableContext bc, List<sObject> records){
+        delete records;
+    }
+    public void finish(Database.BatchableContext bc){
+        System.debug(Constants.TELEMETRIES_DELETED_MESSAGE);
```
Comment (AdrianDlu) [inline InnovativeCars/force-app/main/default/classes/DeleteOldTelemetries.cls]: Deubugi z logging levelem

File: InnovativeCars/force-app/main/default/classes/RenovationTeamRequestSender.cls [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+                    Showroom__c = showroom.Id, 
+                    Origin = Constants.CASE_ORIGIN_WEB
+                    ));
+            }
+
+            Http server = new Http();
+            HttpRequest req = new HttpRequest();
+            req.setEndpoint(Constants.RENOVATION_ENDPOINT);
+            req.setMethod('POST');
+            req.setBody(JSON.serialize(newCases));
```
Comment (AdrianDlu) [inline InnovativeCars/force-app/main/default/classes/RenovationTeamRequestSender.cls]: Do systemu remontowego mialy byc wyslane casy z Id

File: InnovativeCars/force-app/main/default/classes/TestFactory.cls [hunk 12]
```diff
@@ -95,4 +91,23 @@ public class TestFactory {
...
     }
+
+    /**
+    * @author Agnieszka Ząbkowicz
+    * @date 15.06.2021
+    * @description method for delivering telemetries for tests
+    * @param number of telemetries to create and add to database
+    * @return list of created telemetries
+    */
+    public static List<Telemetry__c> createTelemetries(Integer telemetriesNumber){
```
Comment (AdrianDlu) [inline InnovativeCars/force-app/main/default/classes/TestFactory.cls]: parameter name

File: InnovativeCars/force-app/main/default/classes/CarToShowroomHandler.cls [hunk 13]
```diff
@@ -91,4 +88,21 @@ public class CarToShowroomHandler1 {
...
     }
+
+    /**
+    * @author Agnieszka Ząbkowicz
+    * @date 15.06.2021
+    * @description method getting offer's IDs from a list of offers
+    * and sending it to method which sends email
+    * @param newOffers list of offers objects
+    */
+    public static void sendEmailAboutNewOffers(List<Car_To_Showroom__c> newOffers) {
```
Comment (AdrianDlu) [inline InnovativeCars/force-app/main/default/classes/CarToShowroomHandler.cls]: inNewOffers

**Review by AdrianDlu — CHANGES_REQUESTED** [review]

File: InnovativeCars/force-app/main/default/classes/RenovationTeamCaseCreator.cls [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+*/
+public class RenovationTeamCaseCreator implements Queueable, Database.AllowsCallouts{
+    private List<Showroom__c> closedShowrooms;
+    
+    public RenovationTeamCaseCreator(List<Showroom__c> showrooms){
+        this.closedShowrooms = showrooms;
+    }
+
+    public void execute(QueueableContext context){
+        if(!Test.isRunningTest()){
```
Comment (AdrianDlu) [inline InnovativeCars/force-app/main/default/classes/RenovationTeamCaseCreator.cls:25]: Czemu wszystko jest wylaczone z testow? mysle ze tylko linia 26 powina byc w tym ifie

File: InnovativeCars/force-app/main/default/classes/TestFactory.cls (hunk 12 above)
Comment (AdrianDlu) [reply InnovativeCars/force-app/main/default/classes/TestFactory.cls]: #Ignored?

**Review by AdrianDlu — CHANGES_REQUESTED** [review]

**Review by sylwiachr — APPROVED** [review]

**Pull request merged**